src/EditorFeatures/Core/Editor/TextEditApplication.cs

```diff
@@ -17,7 +17,7 @@ internal static void UpdateText(SourceText newText, ITextBuffer buffer, EditOpti
         var oldSnapshot = buffer.CurrentSnapshot;
         var oldText = oldSnapshot.AsText();
         var changes = newText.GetTextChanges(oldText);
-        UpdateText(changes.ToImmutableArray(), buffer, oldSnapshot, oldText, options);
+        UpdateText([.. changes], buffer, oldSnapshot, oldText, options);
```

Comment:
>[.. changes] [](http://example.com/codeflow?start=19&length=12)

Is this smart enough to not allocate?

Reply:
it rewrites it to:

```c#
		int num = 0;
		TextChange[] array = new TextChange[readOnlyList.Count];
		foreach (TextChange item in readOnlyList)
		{
			array[num] = item;
			num++;
		}
		UpdateText(ImmutableCollectionsMarshal.AsImmutableArray(array), buffer, oldSnapshot, oldText, options);
```

which is basically ideal here.

Reply:
Doesn't the old code not allocate since changes is actually an ImmutableArray? 

        public static ImmutableArray<TSource> ToImmutableArray<TSource>(this IEnumerable<TSource> items)
        {
            if (items is ImmutableArray<TSource>)
            {
                return (ImmutableArray<TSource>)items;
            }

            return CreateRange(items);
        }

Reply:
Oh I see. This benefitted from a runtime check.  That's a tougher call. Will have to think about that.

Reply:
Talked with Cyrus about this, and while there is a new allocation here, he's ran perf tests and no regressions appear at that level. Additionally, he's indicated that this particular regression is something that is being investigated to see if it can be addressed at the collection expression level. Not going to block anymore based on this.